File: 资料/协议实现例程/UNO_R3/hu_m40_remote/hu_m40.cpp

/***********************************************************************
 * 
 * 文件名   g_hu_m40.c
 * 作者     hche
 * E-mail   
 * 日期     2023/07/08
 * 版本     V1.1.0
 * 说明     HU遥控手柄解码
 * 
 ************************************************************************
 * 金华市核芯风暴智能科技有限公司
 ***********************************************************************
 
 使用方法：
 
 // 初始化
 void setup() {
   hu_m40_init();
 }
 
 // 轮询
 void loop() {
   if (hu_m40_read() == 1) // ==1有收到数据
   {
     if( hu_m40_analog(HU_LX) > 200 ) // 左摇杆X值> 200
	 {
       // TODO
	 }
	 
     if (hu_m40_button(HU_K1))  // KEY1 按下
     {
       // TODO
     }
   }
 }
 
 */

#include "hu_m40.h"

#include <SPI.h>
#include <nRF24L01.h>
#include <RF24.h>


#define CE_PIN 0
#define CSN_PIN 1
#define PAYLOAD_WIDTH    16

RF24 radio(CE_PIN, CSN_PIN);

const uint8_t HU_M40_ADDR_DEF[5] = "HXFB0";

uint8_t g_hu_m40_rx_data[16];
uint32_t g_hu_m40_last_buttons;
uint32_t g_hu_m40_buttons;

uint8_t g_hu_m40_rx_addr[5] = "HXFB1";

uint8_t* g_hu_m40_tx_addr;

static unsigned long last_read;

uint8_t hu_m40_check_sum(uint8_t* buf, uint8_t len);
void hu_m40_ack(uint8_t* buf, uint8_t len);

uint8_t hu_m40_send_data(uint8_t* buf, uint8_t len);
uint8_t hu_m40_received_data(uint8_t* buf, uint8_t len);


/********************************************************
* 函数名    hu_m40_button
* 输入      button：可选
*          HU_K1
*          HU_K2
*          HU_K3
*          HU_K4
*          HU_K5
*          HU_K6
*           
* 输出      无
* 返回      1：按下
*           0: 松开
* 说明      获取按键状态
********************************************************/
uint8_t hu_m40_button(uint16_t button)
{
  return ((g_hu_m40_buttons & button) > 0);
}

/********************************************************
* 函数名    hu_m40_analog
* 输入      button：可选
*          HU_RX
*          HU_RY
*          HU_LX
*          HU_LY
* 输出      无
* 返回      模拟值
* 说明      获取摇杆状态
********************************************************/
uint8_t hu_m40_analog(uint8_t button)
{
   return g_hu_m40_rx_data[button];
}

/********************************************************
* 函数名    g_hu_m40_init
* 输入      无
* 输出      无
* 返回      1：初始化成功
*           0: 初始化失败
* 说明      初始化
********************************************************/
uint8_t hu_m40_init(void)
{
  uint8_t res;
  uint8_t i;
  
  delay(100);

  radio.begin();
  radio.openWritingPipe(HU_M40_ADDR_DEF);
  radio.openReadingPipe(0, HU_M40_ADDR_DEF);
  radio.setPALevel(RF24_PA_MIN);
  radio.setChannel(25);
  radio.setPayloadSize(PAYLOAD_WIDTH);
  radio.setDataRate(RF24_250KBPS);
  radio.setAutoAck(0, true);
  radio.setRetries(8, 8);
  radio.setCRCLength(RF24_CRC_16);
  radio.startListening();
  g_hu_m40_tx_addr = (uint8_t*)HU_M40_ADDR_DEF;

  last_read = millis();
  
  for( i=0;i<16;i++)
  {
    g_hu_m40_rx_data[i] = 0;
  }

  g_hu_m40_rx_data[HU_RX] = 0x80;
  g_hu_m40_rx_data[HU_RY] = 0x80;
  g_hu_m40_rx_data[HU_LX] = 0x80;
  g_hu_m40_rx_data[HU_LY] = 0x80;

  g_hu_m40_buttons = 0;
  
  return res;
}


/********************************************************
* 函数名    read
* 输入      无
* 输出      无
* 返回      1：成功
*           0: 失败
* 说明      读取
********************************************************/
uint8_t hu_m40_read(void)
{
  uint8_t dat[PAYLOAD_WIDTH]={0};
  uint8_t tmp_buf[PAYLOAD_WIDTH]={0};
  uint8_t i;
  
  unsigned long ticks = millis() - last_read;
  static uint8_t res=1;
  uint8_t flag;
  
  if( ticks > 2000) // 通信超时
  {
    last_read = millis();

    hu_m40_init(); // 重新初始化

    res = 0;
  }

  if( radio.available() )
  {
    radio.read(dat, PAYLOAD_WIDTH);
    if( dat[0] == 1 && dat[1] == 3 && dat[15] == hu_m40_check_sum( dat, 15 ) ) // 校验成功
    {
      
      for( i=0;i<16;i++)
      {
        g_hu_m40_rx_data[i] = dat[i];
      }

      g_hu_m40_buttons = g_hu_m40_rx_data[8];
      
      res = 1;
      
      // 返回数据到遥控器
      dat[0] = 0x01; // 固定为1
      dat[1] = 0x83; // 功能码
      dat[2] = dat[2]; // 序号
      dat[3] = 11; // 长度
      dat[4] = g_hu_m40_rx_addr[0];
      dat[5] = g_hu_m40_rx_addr[1];
      dat[6] = g_hu_m40_rx_addr[2];
      dat[7] = g_hu_m40_rx_addr[3];
      dat[8] = g_hu_m40_rx_addr[4];
      dat[15] = hu_m40_check_sum( dat, 15 );
      
      radio.stopListening();
      
      radio.openWritingPipe(g_hu_m40_tx_addr);
      radio.openReadingPipe(0, g_hu_m40_tx_addr);
      radio.setPALevel(RF24_PA_MIN);
      radio.setChannel(25);
      radio.setPayloadSize(PAYLOAD_WIDTH);
      radio.setDataRate(RF24_250KBPS);
      radio.setAutoAck(0, true);
      radio.setRetries(8, 8);
      radio.setCRCLength(RF24_CRC_16);
      radio.writeBlocking(dat, PAYLOAD_WIDTH, 1000);         //发送数据

      delay(5);
      
      radio.openWritingPipe(g_hu_m40_rx_addr);
      radio.openReadingPipe(0, g_hu_m40_rx_addr);
      radio.setPALevel(RF24_PA_MIN);
      radio.setChannel(25);
      radio.setPayloadSize(PAYLOAD_WIDTH);
      radio.setDataRate(RF24_250KBPS);
      radio.setAutoAck(0, true);
      radio.setRetries(8, 8);
      radio.setCRCLength(RF24_CRC_16);
      radio.startListening();
      g_hu_m40_tx_addr = g_hu_m40_rx_addr;

      last_read = millis();
    }
  }

  return res;
}

uint8_t hu_m40_check_sum(uint8_t* buf, uint8_t len)
{
  uint8_t i, sum;
  sum = 0;
  for( i=0;i<len;i++)
  {
    sum += buf[i];
  }

  return sum;
}

File: 资料/协议实现例程/esp32s3_micropython/esp32s3_micropython/hu_m40.py
#/***********************************************************************
# * 
# * 文件名   hu_m40.py
# * 作者     hche
# * E-mail   
# * 日期     2025/06/30
# * 版本     V0.1.0
# * 说明     HU遥控手柄解码
# * 
# ************************************************************************
# * 金华市核芯风暴智能科技有限公司
# ************************************************************************/

"""HU_M40 driver for MicroPython"""

from machine import Pin, SPI
import nrf24l01
import utime
from micropython import const

HU_K1 = const(0x0001)
HU_K2 = const(0x0002)
HU_K3 = const(0x0004)
HU_K4 = const(0x0008)
HU_K5 = const(0x0010)
HU_K6 = const(0x0020)
HU_RX = const(4)
HU_RY = const(5)
HU_LX = const(6)
HU_LY = const(7)


HU_M40_TIMEOUT_CNT = const(30000)   # 根据不同平台配置通信超时时间，一般配置成2秒
HU_M40_PAYLOAD_WIDTH = const(16)



class HU_M40:

    def __init__(self):

        self.HU_M40_ADDR_DEF = (b"HXFB0")
        self.rx_data = bytearray(16)
        self.last_buttons=0
        self.buttons=0
        self.rx_addr = (b"HXFB1") # 地址码，有冲突时修改此地址
        self.tx_addr=self.HU_M40_ADDR_DEF
        self.ticks=0
        self.nrf=0
    
#/*-------------------------------------------
#
#         以下根据不同平台自行实现  
#         
#----------------------------------------------*/
    
    def init_hw(self):
        CE_PIN = 7
        CSN_PIN = 15
        SCK_PIN = 17
        MOSI_PIN = 16
        MISO_PIN = 18
        
        # 初始化 SPI 接口
        spi = SPI(2, baudrate=1000000, polarity=0, phase=0, sck=Pin(SCK_PIN), mosi=Pin(MOSI_PIN), miso=Pin(MISO_PIN))


        csn = Pin(CSN_PIN, mode=Pin.OUT, value=1)  # Chip Select Not
        ce = Pin(CE_PIN, mode=Pin.OUT, value=0)   # Chip Enable

        # 初始化 NRF24L01
        self.nrf = nrf24l01.NRF24L01(spi, csn, ce, channel=25, payload_size=HU_M40_PAYLOAD_WIDTH)
        self.nrf.reg_write(0x04, (8 << 4) | 8)
        
        
    def tx_mode(self,nrf,txaddr,rxaddr):
        nrf.open_tx_pipe(txaddr)
        nrf.open_rx_pipe(0, rxaddr)
        
    def rx_mode(self,nrf,txaddr,rxaddr):
        nrf.open_tx_pipe(txaddr)
        nrf.open_rx_pipe(0, rxaddr)
        nrf.start_listening()

    def send_data(self,nrf,buf,length):
        nrf.send(buf,200)
        #nrf.send_start(buf)
        #start = utime.ticks_ms()
        #result = None
        #while result is None and utime.ticks_diff(utime.ticks_ms(), start) < 100:
        #    result = nrf.send_done()  # 1 == success, 2 == fail


    def received_data(self,nrf,buf,length):
        if nrf.any():
            buf1=nrf.recv() # 接收数据
            nrf.stop_listening()
            for i in range(16):
                buf[i] = buf1[i]
            
            return 1
        return 0
    
#/*-------------------------------------------
#
#         一般情况下以下不用修改
#         
#----------------------------------------------*/
    def button(self, button):
        return ((self.buttons & button) > 0);
        
    def analog(self, button):
        return self.rx_data[button]
        
    def check_sum(self,buf, length):
        sum=0
        for i in range(length):
            sum += buf[i]
        
        return sum&0xFF
        
        
    def init(self):
        #print("hu_m40_init")
        self.init_hw()
        nrf = self.nrf
        
        self.rx_mode(nrf,self.HU_M40_ADDR_DEF,self.HU_M40_ADDR_DEF)
        self.tx_addr=self.HU_M40_ADDR_DEF
        
        for i in range(16):
            self.rx_data[i]=0
        
        self.rx_data[HU_RX] = 0x80;
        self.rx_data[HU_RY] = 0x80;
        self.rx_data[HU_LX] = 0x80;
        self.rx_data[HU_LY] = 0x80;
        
        self.buttons = 0;
        
        
        
    def read(self):
        
        nrf = self.nrf
        res = 0
        dat = bytearray(16)
        
        
        self.ticks+=1
        
        if self.ticks > HU_M40_TIMEOUT_CNT:  # 通信超时
        
            self.ticks = 0;

            self.init() # 重新初始化

            res = 0
        

        if self.received_data( self.nrf, dat, HU_M40_PAYLOAD_WIDTH ) :
            #print("收到数据:", dat)
            #print("dat[0] == 1:", dat[0] == 1)
            #print("dat[1] == 3:", dat[1] == 3)
            #print("dat[15]:", dat[15])
            #print("self.check_sum( dat, 15 ):", self.check_sum( dat, 15 ))
            
            if dat[0] == 1 and dat[1] == 3 and dat[15] == self.check_sum( dat, 15 ) : # 校验成功
            
              
                  for i in range(16):
                  
                      self.rx_data[i] = dat[i];
                  

                  self.buttons = self.rx_data[8];
                  
                  res = 1;
                  self.ticks = 0;
                  
                  # 返回数据到遥控器
                  dat[0] = 0x01; # 固定为1
                  dat[1] = 0x83; # 功能码
                  dat[2] = dat[2]; # 序号
                  dat[3] = 11; # 长度
                  dat[4] = self.rx_addr[0];
                  dat[5] = self.rx_addr[1];
                  dat[6] = self.rx_addr[2];
                  dat[7] = self.rx_addr[3];
                  dat[8] = self.rx_addr[4];
                  dat[15] = self.check_sum( dat, 15 );
                  
                  #print("tx_addr:", self.tx_addr)
                  self.tx_mode( nrf, self.tx_addr, self.tx_addr );
                  self.send_data(nrf, dat, HU_M40_PAYLOAD_WIDTH);         #发送数据
                  #print("发送数据:", dat)
                  
                  
                  self.rx_mode( nrf, self.rx_addr, self.rx_addr );
                  self.tx_addr = self.rx_addr;
        return res




__version__ = '0.1.0'


